File: backend/app/monte_carlo.py
import random
import numpy as np


def simulate_match(xg_home, xg_away):

    home_goals = np.random.poisson(xg_home)
    away_goals = np.random.poisson(xg_away)

    return home_goals, away_goals


def monte_carlo_match(xg_home, xg_away, simulations=20000):

    home_wins = 0
    draws = 0
    away_wins = 0

    for _ in range(simulations):

        hg, ag = simulate_match(xg_home, xg_away)

        if hg > ag:
            home_wins += 1
        elif hg == ag:
            draws += 1
        else:
            away_wins += 1

    return {
        "home_win": home_wins / simulations,
        "draw": draws / simulations,
        "away_win": away_wins / simulations
    }

File: frontend/src/analytics/montecarlo.js
export function simulateMatch(xgHome,xgAway,simulations=10000){

  function poisson(lambda){
    let L=Math.exp(-lambda)
    let p=1
    let k=0

    while(p>L){
      k++
      p*=Math.random()
    }

    return k-1
  }

  let homeWins=0
  let draws=0
  let awayWins=0

  for(let i=0;i<simulations;i++){

    let hg=poisson(xgHome)
    let ag=poisson(xgAway)

    if(hg>ag) homeWins++
    else if(hg===ag) draws++
    else awayWins++

  }

  return{
    home:homeWins/simulations,
    draw:draws/simulations,
    away:awayWins/simulations
  }

}